pkg/kv/kvserver/replica_range_lease_test.go

```diff
@@ -0,0 +1,150 @@
+		ProposedTS: &hlc.ClockTimestamp{WallTime: ts[1].WallTime - 100},
+	}
+	epoLease := roachpb.Lease{
+		Replica:    expLease.Replica,
+		Start:      expLease.Start,
+		Epoch:      5,
+		ProposedTS: expLease.ProposedTS,
+	}
+
+	oldLiveness := livenesspb.Liveness{
+		NodeID: 1, Epoch: 4, Expiration: hlc.LegacyTimestamp{WallTime: ts[1].WallTime},
+	}
+	curLiveness := livenesspb.Liveness{
+		NodeID: 1, Epoch: 5, Expiration: hlc.LegacyTimestamp{WallTime: ts[4].WallTime},
+	}
+	expLiveness := livenesspb.Liveness{
+		NodeID: 1, Epoch: 6, Expiration: hlc.LegacyTimestamp{WallTime: ts[6].WallTime},
+	}
+
+	for _, tc := range []struct {
+		lease         roachpb.Lease
+		now           hlc.ClockTimestamp
+		minProposedTS hlc.ClockTimestamp
+		reqTS         hlc.Timestamp
+		liveness      livenesspb.Liveness
+		want          kvserverpb.LeaseState
+		wantErr       string
+	}{
+		// Expiration-based lease, EXPIRED.
+		{lease: expLease, now: ts[5], want: kvserverpb.LeaseState_EXPIRED},
+		{lease: expLease, now: ts[4], want: kvserverpb.LeaseState_EXPIRED},
+		// Expiration-based lease, PROSCRIBED.
+		{lease: expLease, now: ts[3], minProposedTS: ts[1], want: kvserverpb.LeaseState_PROSCRIBED},
+		{lease: expLease, now: ts[3], minProposedTS: ts[2], want: kvserverpb.LeaseState_PROSCRIBED},
+		{lease: expLease, now: ts[3], minProposedTS: ts[3], reqTS: inStasis,
+			want: kvserverpb.LeaseState_PROSCRIBED},
+		{lease: expLease, now: ts[3], minProposedTS: ts[4], reqTS: inStasis,
+			want: kvserverpb.LeaseState_PROSCRIBED},
+		// Expiration-based lease, UNUSABLE.
+		{lease: expLease, now: ts[3], reqTS: inStasis, want: kvserverpb.LeaseState_UNUSABLE},
+		// Expiration-based lease, VALID.
+		{lease: expLease, now: ts[2], reqTS: ts[2].ToTimestamp(), want: kvserverpb.LeaseState_VALID},
+
+		// Epoch-based lease, EXPIRED.
+		{lease: epoLease, now: ts[5], liveness: curLiveness, want: kvserverpb.LeaseState_EXPIRED},
+		{lease: epoLease, now: ts[4], liveness: curLiveness, want: kvserverpb.LeaseState_EXPIRED},
+		{lease: epoLease, now: ts[2], liveness: expLiveness, want: kvserverpb.LeaseState_EXPIRED},
+		// Epoch-based lease, PROSCRIBED.
+		{lease: epoLease, now: ts[3], liveness: curLiveness, minProposedTS: ts[1],
+			want: kvserverpb.LeaseState_PROSCRIBED},
+		{lease: epoLease, now: ts[2], liveness: curLiveness, minProposedTS: ts[2],
+			want: kvserverpb.LeaseState_PROSCRIBED},
+		// Epoch-based lease, UNUSABLE.
+		{lease: epoLease, now: ts[3], liveness: curLiveness, reqTS: ts[3].ToTimestamp(),
+			want: kvserverpb.LeaseState_UNUSABLE},
+		// Epoch-based lease, VALID.
+		{lease: epoLease, now: ts[2], liveness: curLiveness, want: kvserverpb.LeaseState_VALID},
+
+		// Epoch-based lease, ERROR.
+		{lease: epoLease, now: ts[2], want: kvserverpb.LeaseState_ERROR,
+			wantErr: "liveness record not found"},
+		{lease: epoLease, now: ts[2], liveness: oldLiveness, want: kvserverpb.LeaseState_ERROR,
+			wantErr: "node liveness info for n1 is stale"},
+	} {
+		t.Run("", func(t *testing.T) {
```

Comment:
nit: Instead of comments above the test cases, consider adding a `desc` field and use that as the test name too. It's annoying to debug failures when you have to count to find the correct case.